![Alt Text](http://spark-mooc.github.io/web-assets/images/python-logo-master-v3-TM-flattened_small.png)

![Alt Text](http://spark-mooc.github.io/web-assets/images/ta_Spark-logo-small.png)

#### question 2

first : by copying the lines of code it shows us that it has a many kinds of errors

In [1]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.pipeline import pipeline
from sklearn_linear_model import SGDClassifier

categories = [ 'rec.sport.baseball','rec.sport.hockey']
twenty_train = fetch_20newsgroups(subset='test', categories=categories)
twenty_test = fetch_20newsgroups(subset='train', categories=categories, shuffle=True, random_state=42)

text_clf = Pipeline([('vect', CountVectorizer()),
                     ('tfidf', TfidfTransformer()),
                     ('clf', SGDClassifier(loss='hinge', penalty='l2',
                                           alpha=1e-3, random_state=42)),
])
text_clf.fit(twenty_train.data, twenty_test.target)  
predicted = text_clf.predict(twenty_test.data)

ImportError: cannot import name 'pipeline'

#### the first error is that the (p) in pipelines should be in capital form

In [2]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.pipeline import Pipeline
from sklearn_linear_model import SGDClassifier

categories = [ 'rec.sport.baseball','rec.sport.hockey']
twenty_train = fetch_20newsgroups(subset='test', categories=categories)
twenty_test = fetch_20newsgroups(subset='train', categories=categories, shuffle=True, random_state=42)

text_clf = Pipeline([('vect', CountVectorizer()),
                     ('tfidf', TfidfTransformer()),
                     ('clf', SGDClassifier(loss='hinge', penalty='l2',
                                           alpha=1e-3, random_state=42)),
])
text_clf.fit(twenty_train.data, twenty_test.target)  
predicted = text_clf.predict(twenty_test.data)

ModuleNotFoundError: No module named 'sklearn_linear_model'

#### 2- then we have to include that we imported the correct libaries

-import numpy as np -from sklearn import datasets, linear_model

In [3]:
import numpy as np
from sklearn import datasets, linear_model
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.pipeline import Pipeline
from sklearn_linear_model import SGDClassifier

categories = [ 'rec.sport.baseball','rec.sport.hockey']
twenty_train = fetch_20newsgroups(subset='test', categories=categories)
twenty_test = fetch_20newsgroups(subset='train', categories=categories, shuffle=True, random_state=42)

text_clf = Pipeline([('vect', CountVectorizer()),
                     ('tfidf', TfidfTransformer()),
                     ('clf', SGDClassifier(loss='hinge', penalty='l2',
                                           alpha=1e-3, random_state=42)),
])
text_clf.fit(twenty_train.data, twenty_test.target)  
predicted = text_clf.predict(twenty_test.data)

ModuleNotFoundError: No module named 'sklearn_linear_model'

#### 3- (correcting sklearn_linear_model ) to (sklearn.linear_model )

In [5]:
import numpy as np
from sklearn import datasets, linear_model
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import SGDClassifier

categories = ['rec.sport.baseball','rec.sport.hockey']
twenty_train = fetch_20newsgroups(subset='test', categories=categories)
twenty_test = fetch_20newsgroups(subset='train', categories=categories, shuffle=True, random_state=42)

text_clf = Pipeline([('vect', CountVectorizer()),
                     ('tfidf', TfidfTransformer()),
                     ('clf', SGDClassifier(loss='hinge', penalty='l2',
                                           alpha=1e-3, random_state=42)),
])
text_clf.fit(twenty_train.data, twenty_test.target)  
predicted = text_clf.predict(twenty_test.data)

ValueError: Found input variables with inconsistent numbers of samples: [796, 1197]

#### 4-there is an error also in lines 8 and 9 we have to change test by train and train by test to be a readable code

In [6]:
import numpy as np
from sklearn import datasets, linear_model
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import SGDClassifier

categories = ['rec.sport.baseball','rec.sport.hockey']
twenty_train = fetch_20newsgroups(subset='train', categories=categories)
twenty_test = fetch_20newsgroups(subset='test', categories=categories, shuffle=True, random_state=42)

text_clf = Pipeline([('vect', CountVectorizer()),
                     ('tfidf', TfidfTransformer()),
                     ('clf', SGDClassifier(loss='hinge', penalty='l2',
                                           alpha=1e-3, random_state=42)),])
text_clf.fit(twenty_train.data, twenty_test.target)  
predicted = text_clf.predict(twenty_test.data)

ValueError: Found input variables with inconsistent numbers of samples: [1197, 796]

#### 5- the last two errors are in the extra comma after the brackets in the 12th line.
#### and in the 13 you have to change the (test) with (train)



In [7]:
import numpy as np
from sklearn import datasets, linear_model
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import SGDClassifier

categories = ['rec.sport.baseball','rec.sport.hockey']
twenty_train = fetch_20newsgroups(subset='test', categories=categories)
twenty_test = fetch_20newsgroups(subset='train', categories=categories, shuffle=True, random_state=42)

text_clf = Pipeline([('vect', CountVectorizer()),
                     ('tfidf', TfidfTransformer()),
                     ('clf', SGDClassifier(loss='hinge', penalty='l2',
                                           alpha=1e-3, random_state=42))])
text_clf.fit(twenty_train.data, twenty_train.target)  
predicted = text_clf.predict(twenty_test.data)


#### - now it can be read

#### How many observations are in the training dataset? , How many features are in the training dataset?

In [8]:
count_vect = CountVectorizer()
X_train_counts = count_vect.fit_transform(twenty_train.data)
X_train_counts.shape

(796, 15074)

#### here are 796 observations and 15074 features in the dataset

#### How well did your model perform?

In [9]:
from sklearn import metrics
print(metrics.classification_report(twenty_test.target, predicted, target_names = twenty_test.target_names))

                    precision    recall  f1-score   support

rec.sport.baseball       0.95      0.98      0.97       597
  rec.sport.hockey       0.98      0.95      0.97       600

       avg / total       0.97      0.97      0.97      1197



#### the model is preforming well with the precision of 0.97